### ISSUE TITLE
[QUESTION] Parallelizing downloading following lists hangs in the terminal. Is there a way to parallelize downloading anything in Twint?

### ISSUE THREAD

**Issue by youssefavx** [twintproject/twint#809]
- [] Python version is 3.7.4;
- [] Updated Twint with `pip3 install --user --upgrade -e git+https://github.com/twintproject/twint.git@origin/master#egg=twint`;
- [] I have searched the issues and there are no duplicates of this issue/question/request.

### Command Ran

In one python file I had this code: 

```
from numba import jit 
import twint

def downloadfollowingsingleuser(utarget):

    if os.path.exists('data/' + str(utarget) + ' resume following.csv'):
        os.remove('data/' + str(utarget) + ' resume following.csv')
    print("Downloading who " + str(utarget) + " is following")
    y = twint.Config()
    y.Username = str(utarget.lower())
    y.Store_csv = True
    #y.Limit = 30000
    y.Resume = "data/" + str(utarget) + " resume following.csv"

    #y.Output = str(utarget) + " following.csv"
    y.Output = "data/" + str(utarget) + " following.csv"

    while True:
        try:
            twint.run.Following(y)
            break
        except aiohttp.ClientConnectorError:
            time.sleep(1)
            print('Cant connect to internet. Restarting...')

        except aiohttp.ClientOSError:
            time.sleep(1)
            print('Cant connect to internet. Restarting...')

        except aiohttp.ServerDisconnectedError:
            time.sleep(1)
            print('Cant connect to internet. Restarting...')

        except asyncio.TimeoutError:
            time.sleep(1)
            print('Cant connect to internet. Restarting...')

        except KeyboardInterrupt:
            os.remove("data/" + str(utarget) + " following.csv")
            os.remove("data/" + str(utarget) + " resume following.csv")
            raise ValueError("Quit program. Deleted following and resume files for " + str(utarget) + ".")


@jit(parallel=True)
def downloadfollowingparsl(userlist):
    for user in userlist:
        downloadfollowingsingleuser(user)
```

In another python file, I imported this code, and used it like so: 

`downloadfollowing(targetuserfollowing)`

`targetuserfollowing` is a list of users a given user is following.

### Description of Issue

I should note that I tried parallelizing this code using threading, using parsl, numba jit, and using joblib. Even though the function was successful in downloading the files, right after it downloads the files (if this is run in a terminal) for some odd reason, it simply freezes, or ends the program. It continues to the next line but at some point it simply stops (for instance it does not go fully through a for loop of 90 elements but only the first 2 elements). This only happens when running it in the terminal.

Searching online a bit, some people have suggested that trying to parallelize code that is already using some form of parallelization might lead to this problem. Still I have no idea how to solve this. 

An important note is: On jupyter notebook, this code works fine. If you separate it into chunks and then run them separately, you simply have to run this code block a second time for it to work (running it the first time it does not work): 

```
for userone in targetuserfollowing:
    print("Getting following list for " + userone)
    useronefollowing = getfollowinglist(userone)
    for usertwo in useronefollowing:
        if usertwo in sourceuserfollowers:
            print('Found', usertwo, 'in followers')
```

getfollowinglist is a function that searches through a directory to find a file like this: 'username following.csv' and appends the users from the csv to a list. This gives us who userone is following. 

sourceuserfollowers is a list of a given users followers.

### Environment Details
>Mac OS Mojave 
> Terminal

**Comment by youssefavx** [comment]
Here is parallelization using Parsl (just in case that is also of interest): 


```
#Parsl download following list
import parsl
from parsl import python_app

parsl.load()

@python_app
def downloadfollowingsingleuserparsl(utarget):
    if os.path.exists('data/' + str(utarget) + ' resume following.csv'):
        os.remove('data/' + str(utarget) + ' resume following.csv')
    print("Downloading who " + str(utarget) + " is following")
    y = twint.Config()
    y.Username = str(utarget.lower())
    y.Store_csv = True
    #y.Limit = 30000
    y.Resume = "data/" + str(utarget) + " resume following.csv"

    #y.Output = str(utarget) + " following.csv"
    y.Output = "data/" + str(utarget) + " following.csv"

    while True:
        try:
            twint.run.Following(y)
            break
        except aiohttp.ClientConnectorError:
            time.sleep(1)
            print('Cant connect to internet. Restarting...')
            #twint.run.Followers(x)
        except aiohttp.ClientOSError:
            time.sleep(1)
            print('Cant connect to internet. Restarting...')
            #twint.run.Followers(x)
        except aiohttp.ServerDisconnectedError:
            time.sleep(1)
            print('Cant connect to internet. Restarting...')
            #twint.run.Followers(x)
        except asyncio.TimeoutError:
            time.sleep(1)
            print('Cant connect to internet. Restarting...')
        except KeyboardInterrupt:
            os.remove("data/" + str(utarget) + " following.csv")
            os.remove("data/" + str(utarget) + " resume following.csv")
            raise ValueError("Quit program. Deleted following and resume files for " + str(utarget) + ".")


def downloadfollowingparsl(userlist):
    for user in userlist:
        downloadfollowingsingleuserparsl(user)
```

**Comment by youssefavx** [comment]
And here is using the threading module: 

```
import twint
import threading
import asyncio 
import os 

def downloadfollowingforsingleuserparallel(utarget):
    asyncio.set_event_loop(asyncio.new_event_loop())
    if not os.path.exists("data/" + str(utarget) + " following.csv"):

        if os.path.exists('data/' + str(utarget) + ' resume following.csv'):
            os.remove('data/' + str(utarget) + ' resume following.csv')
        print("Downloading who " + str(utarget) + " is following")
        y = twint.Config()
        y.Username = str(utarget.lower())
        y.Store_csv = True
        #y.Limit = 30000
        y.Resume = "data/" + str(utarget) + " resume following.csv"

        #y.Output = str(utarget) + " following.csv"
        y.Output = "data/" + str(utarget) + " following.csv"

        while True:
            try:
                twint.run.Following(y)
                break
            except aiohttp.ClientConnectorError:
                time.sleep(1)
                print('Cant connect to internet. Restarting...')
                #twint.run.Followers(x)
            except aiohttp.ClientOSError:
                time.sleep(1)
                print('Cant connect to internet. Restarting...')
                #twint.run.Followers(x)
            except aiohttp.ServerDisconnectedError:
                time.sleep(1)
                print('Cant connect to internet. Restarting...')
                #twint.run.Followers(x)
            except asyncio.TimeoutError:
                time.sleep(1)
                print('Cant connect to internet. Restarting...')
            except KeyboardInterrupt:
                os.remove("data/" + str(utarget) + " following.csv")
                os.remove("data/" + str(utarget) + " resume following.csv")
                raise ValueError("Quit program. Deleted following and resume files for " + str(utarget) + ".")




def downloadfollowing(users):
    while True:
        if len(users) > 0:
            user_to_be_processed = users.pop()
            threading.Thread(target = downloadfollowingforsingleuserparallel,args=(user_to_be_processed,)).start()
```

**Comment by youssefavx** [comment]
When I try using those functions (e.g. downloadfollowing(usernames)), and then do some further operations after with the information like getting the data and looping through it, I find that the program hangs in terminal.

**Comment by youssefavx** [comment]
I also should note that before this download following lists function, I also have a function that downloads following number counts in parallel. But since the download following lists function works after that function I'm assuming nothing is wrong with that function.

**Comment by hockeybro12** [comment]
@youssefavx did you ever manage to solve this?
